File: lib/models/external_delivery_token.dart
import 'dart:convert';
import 'dart:math';
import 'package:cloud_firestore/cloud_firestore.dart';

class ExternalDeliveryToken {
  final String id;
  final String deliveryId;
  final String companyId;
  final String providerName;
  final String? providerContact;
  final DateTime createdAt;
  final DateTime? expiresAt;
  final bool isUsed;
  final List<String> uploadedFiles;
  final DateTime? uploadedAt;
  final String? uploadNotes;

  ExternalDeliveryToken({
    required this.id,
    required this.deliveryId,
    required this.companyId,
    required this.providerName,
    this.providerContact,
    required this.createdAt,
    this.expiresAt,
    this.isUsed = false,
    this.uploadedFiles = const [],
    this.uploadedAt,
    this.uploadNotes,
  });

  /// Generates a secure random token
  static String generateToken() {
    final random = Random.secure();
    final bytes = List<int>.generate(24, (i) => random.nextInt(256));
    return base64Url
        .encode(bytes)
        .replaceAll(RegExp(r'[^a-zA-Z0-9]'), '')
        .substring(0, 32);
  }

  /// Checks if token is valid for use
  bool isValid() {
    if (isUsed) return false;
    if (expiresAt != null && DateTime.now().isAfter(expiresAt!)) {
      return false;
    }
    return true;
  }

  factory ExternalDeliveryToken.fromFirestore(DocumentSnapshot doc) {
    Map<String, dynamic> data = doc.data() as Map<String, dynamic>;

    return ExternalDeliveryToken(
      id: doc.id,
      deliveryId: data['deliveryId'] ?? '',
      companyId: data['companyId'] ?? '',
      providerName: data['providerName'] ?? '',
      providerContact: data['providerContact'],
      createdAt: (data['createdAt'] as Timestamp).toDate(),
      expiresAt: data['expiresAt'] != null
          ? (data['expiresAt'] as Timestamp).toDate()
          : null,
      isUsed: data['isUsed'] ?? false,
      uploadedFiles:
          (data['uploadedFiles'] as List<dynamic>?)?.cast<String>() ?? [],
      uploadedAt: data['uploadedAt'] != null
          ? (data['uploadedAt'] as Timestamp).toDate()
          : null,
      uploadNotes: data['uploadNotes'],
    );
  }

  Map<String, dynamic> toFirestore() {
    return {
      'deliveryId': deliveryId,
      'companyId': companyId,
      'providerName': providerName,
      'providerContact': providerContact,
      'createdAt': Timestamp.fromDate(createdAt),
      if (expiresAt != null) 'expiresAt': Timestamp.fromDate(expiresAt!),
      'isUsed': isUsed,
      'uploadedFiles': uploadedFiles,
      if (uploadedAt != null) 'uploadedAt': Timestamp.fromDate(uploadedAt!),
      'uploadNotes': uploadNotes,
    };
  }

  ExternalDeliveryToken copyWith({
    String? id,
    String? deliveryId,
    String? companyId,
    String? providerName,
    String? providerContact,
    DateTime? createdAt,
    DateTime? expiresAt,
    bool? isUsed,
    List<String>? uploadedFiles,
    DateTime? uploadedAt,
    String? uploadNotes,
  }) {
    return ExternalDeliveryToken(
      id: id ?? this.id,
      deliveryId: deliveryId ?? this.deliveryId,
      companyId: companyId ?? this.companyId,
      providerName: providerName ?? this.providerName,
      providerContact: providerContact ?? this.providerContact,
      createdAt: createdAt ?? this.createdAt,
      expiresAt: expiresAt ?? this.expiresAt,
      isUsed: isUsed ?? this.isUsed,
      uploadedFiles: uploadedFiles ?? this.uploadedFiles,
      uploadedAt: uploadedAt ?? this.uploadedAt,
      uploadNotes: uploadNotes ?? this.uploadNotes,
    );
  }
}

File: mobile-rn/src/models/externalDeliveryToken.ts
import firestore, { FirebaseFirestoreTypes } from '@react-native-firebase/firestore';
import SHA256 from 'crypto-js/sha256';

/**
 * Ported from lib/models/external_delivery_token.dart (verified against source on
 * 2026-06-23) — scoped to what create_delivery_screen.dart actually needs
 * (createUploadToken/generateUploadLink). `getToken`/`uploadDocuments`/
 * `uploadDocumentsWeb`/`revokeToken`/`getDeliveryForToken` from
 * external_upload_service.dart are NOT ported yet — those back the actual upload flow
 * for lib/screens/external/upload_screen.dart (`/upload/:token`), a separate
 * not-yet-built public route, not this admin screen.
 *
 * Deviation: token generation uses crypto-js's SHA256 instead of Dart's
 * `Random.secure()` + base64Url, matching the precedent already set by
 * podTokenRepository.ts's generateToken() for the near-identical PODAccessToken case.
 */
export interface ExternalDeliveryToken {
  id: string;
  deliveryId: string;
  companyId: string;
  providerName: string;
  providerContact?: string;
  createdAt: Date;
  expiresAt?: Date;
  isUsed: boolean;
  uploadedFiles: string[];
  uploadedAt?: Date;
  uploadNotes?: string;
}

/** Mirrors ExternalDeliveryToken.generateToken(). */
export function generateExternalUploadToken(deliveryId: string): string {
  const timestamp = Date.now();
  const random = Math.floor(Math.random() * 1_000_000_000);
  return SHA256(`${deliveryId}-${timestamp}-${random}`).toString().substring(0, 32);
}

/** Mirrors ExternalDeliveryToken.isValid(). */
export function isExternalUploadTokenValid(token: ExternalDeliveryToken): boolean {
  if (token.isUsed) return false;
  if (token.expiresAt && new Date() > token.expiresAt) return false;
  return true;
}

export function externalDeliveryTokenFromFirestore(doc: FirebaseFirestoreTypes.DocumentSnapshot): ExternalDeliveryToken {
  const data = doc.data() ?? {};
  return {
    id: doc.id,
    deliveryId: data.deliveryId ?? '',
    companyId: data.companyId ?? '',
    providerName: data.providerName ?? '',
    providerContact: data.providerContact ?? undefined,
    createdAt: (data.createdAt as FirebaseFirestoreTypes.Timestamp)?.toDate() ?? new Date(),
    expiresAt: (data.expiresAt as FirebaseFirestoreTypes.Timestamp | undefined)?.toDate(),
    isUsed: data.isUsed ?? false,
    uploadedFiles: Array.isArray(data.uploadedFiles) ? data.uploadedFiles : [],
    uploadedAt: (data.uploadedAt as FirebaseFirestoreTypes.Timestamp | undefined)?.toDate(),
    uploadNotes: data.uploadNotes ?? undefined,
  };
}

export function externalDeliveryTokenToFirestore(token: Omit<ExternalDeliveryToken, 'id'>): Record<string, unknown> {
  return {
    deliveryId: token.deliveryId,
    companyId: token.companyId,
    providerName: token.providerName,
    providerContact: token.providerContact ?? null,
    createdAt: firestore.Timestamp.fromDate(token.createdAt),
    expiresAt: token.expiresAt ? firestore.Timestamp.fromDate(token.expiresAt) : null,
    isUsed: token.isUsed,
    uploadedFiles: token.uploadedFiles,
    uploadedAt: token.uploadedAt ? firestore.Timestamp.fromDate(token.uploadedAt) : null,
    uploadNotes: token.uploadNotes ?? null,
  };
}
